GH-35409: [Python][Docs] Clarify S3FileSystem Credentials chain for EC2

kevinjqliu opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->
When resolving AWS credentials on EC2 hosts, the underlying AWS SDK also looks at the EC2 Instance Metadata Service. 

I want to document this behavior for `pyarrow`.  The [`s3fs` documentation](https://s3fs.readthedocs.io/en/latest/#credentials) mention this specific case for EC2.

### What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->
Documentation for the behavior described above. 

#### Technical Details
`S3FileSystem` uses the [`CS3Options.Defaults()`](https://github.com/apache/arrow/blob/5de56928e0fe43f02005552eee058de57ffb2682/python/pyarrow/_s3fs.pyx#L317) option when no credentials are passed into the constructor.  It utilizes the [`Aws::Auth::DefaultAWSCredentialsProviderChain`](https://github.com/apache/arrow/blob/1de159d0f6763766c19b183dd309b8757723b43a/cpp/src/arrow/filesystem/s3fs.cc#L213)

The C++ implementation of [`DefaultAWSCredentialsProviderChain`](https://sdk.amazonaws.com/cpp/api/0.14.3/class_aws_1_1_auth_1_1_default_a_w_s_credentials_provider_chain.html) not only [reads the environment variable](https://sdk.amazonaws.com/cpp/api/0.14.3/class_aws_1_1_auth_1_1_environment_a_w_s_credentials_provider.html) when trying to resolve AWS credentials, but also [looks at profile config](https://sdk.amazonaws.com/cpp/api/0.14.3/class_aws_1_1_auth_1_1_profile_config_file_a_w_s_credentials_provider.html) and the [EC2 Instance Metadata Service](https://sdk.amazonaws.com/cpp/api/0.14.3/class_aws_1_1_auth_1_1_instance_profile_credentials_provider.html). 


### Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->
No, just documentation changes

### Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->
Yes, changing public documentation

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->
* Closes: #35409


### Render Changes
Render the changes locally via [Building the doc](https://arrow.apache.org/docs/developers/documentation.html#building-docs): 
`docs/source/python/filesystems.rst`:
![Screenshot 2023-07-30 at 6 22 02 PM](https://github.com/apache/arrow/assets/9057843/6af053a3-e7a7-4a68-a5b5-02c50e9290c6)

`python/pyarrow/_s3fs.pyx`:
![Screenshot 2023-07-31 at 3 31 30 PM](https://github.com/apache/arrow/assets/9057843/d79768be-67ce-46c0-88ed-a833e540f77d)

jorisvandenbossche commented:
@kevinjqliu thanks for the contribution!

We have a section about S3FileSystem in the user guide as well: https://arrow.apache.org/docs/dev/python/filesystems.html#s3. Could you have a look to see if that might require an equivalent update?

jorisvandenbossche commented on python/pyarrow/_s3fs.pyx:
```diff
@@ -140,13 +140,17 @@ cdef class S3FileSystem(FileSystem):
     specified role.
+    
+    If neither access_key nor secret_key are provided, and role_arn is also not
+    provided, then attempts to initialize from the Default AWS Credentials Provider Chain in order of:
```
Formatting nit: can you limit the line length here?

jorisvandenbossche commented on python/pyarrow/_s3fs.pyx:
```diff
@@ -140,13 +140,17 @@ cdef class S3FileSystem(FileSystem):
+    
+    If neither access_key nor secret_key are provided, and role_arn is also not
+    provided, then attempts to initialize from the Default AWS Credentials Provider Chain in order of:
+    * `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN` environment variables
```
```suggestion

    * `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN` environment variables
```

RestructuredText markup needs a blank line before a bullet point list

kevinjqliu commented:
@jorisvandenbossche fixed a formatting issue. there was something wrong with CI too, hence the force pushes.

pitrou commented:
@jorisvandenbossche Would you like to take another look?

kevinjqliu commented:
hey, @jorisvandenbossche wanted to ping on this again. thanks!

kou commented:
@github-actions crossbow submit docs-preview

kou commented on python/pyarrow/_s3fs.pyx:
```diff
@@ -140,14 +140,20 @@ cdef class S3FileSystem(FileSystem):
+    provided, then attempts establish the credentials automatically.
+    S3FileSystem will try the following methods, in order:
+
+    * `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN` environment variables
+    * configuration files such as `~/.aws/credentials` and `~/.aws/config`
```
We need two `` ` ``s for reStructuredText:

```suggestion
    * ``AWS_ACCESS_KEY_ID``, ``AWS_SECRET_ACCESS_KEY``, and ``AWS_SESSION_TOKEN`` environment variables
    * configuration files such as ``~/.aws/credentials`` and ``~/.aws/config``
```

kou commented on python/pyarrow/_s3fs.pyx:
```diff
@@ -140,14 +140,20 @@ cdef class S3FileSystem(FileSystem):
     specified role.
 
+    If neither access_key nor secret_key are provided, and role_arn is also not
+    provided, then attempts establish the credentials automatically.
```
Is `to` missing?

```suggestion
    provided, then attempts to establish the credentials automatically.
```

kou commented:
@github-actions crossbow submit preview-docs

kevinjqliu commented:
thanks @kou I've applied your suggestions

kou commented:
@github-actions crossbow submit preview-docs

kevinjqliu commented:
@kou built the changes locally, I've attached a screenshot above

kou commented:
Thanks but we have 2 changed pages but the screenshot shows only 1 page.

kevinjqliu commented:
The visible change is from `docs/source/python/filesystems.rst`. 
`python/pyarrow/_s3fs.pyx` is a code comment change.

kou commented:
We use it for API reference: https://arrow.apache.org/docs/python/generated/pyarrow.fs.S3FileSystem.html#pyarrow.fs.S3FileSystem

kevinjqliu commented:
> We use it for API reference: https://arrow.apache.org/docs/python/generated/pyarrow.fs.S3FileSystem.html#pyarrow.fs.S3FileSystem

I see. I'm following [Building a single directory for dev purposes without all the pre-requisites](https://arrow.apache.org/docs/developers/documentation.html#building-a-single-directory-for-dev-purposes-without-all-the-pre-requisites) to render the changes for `docs/source/python`.

The changes in `python/pyarrow/_s3fs.pyx` does not show up. How can I render that part?

kou commented:
We need pyarrow to build documents in `python/pyarrow/_s3fs.pyx`. So we can't use the approach.

We need to use the full (at least including pyarrow) build approach.
https://arrow.apache.org/docs/developers/documentation.html#building-with-docker is easier than https://arrow.apache.org/docs/developers/documentation.html#building .

kou commented:
Could you rebase on main to use #36943 ?

kevinjqliu commented:
@kou I've rebased the branch and was able to build it locally using docker. Updated the description above to render changes to both files. 

Thanks for guiding me through the process

kou commented:
@github-actions crossbow submit preview-docs

kou commented:
http://crossbow.voltrondata.com/pr_docs/35312/

* http://crossbow.voltrondata.com/pr_docs/35312/python/filesystems.html#s3
* http://crossbow.voltrondata.com/pr_docs/35312/python/generated/pyarrow.fs.S3FileSystem.html

kou reviewed (approved):
+1
